### PR TITLE
Scikit-Learn compatible rewrite

### PR DESCRIPTION
**Pull request by ilias-karimalis**

### CODE REVIEW

File: .gitignore
Comment (HaydenMcT) [inline .gitignore:1]: We should leave this file, yes?

File: LICENSE [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
-   this list of conditions and the following disclaimer in the documentation
-   and/or other materials provided with the distribution.
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
 
-3. Neither the name of the copyright holder nor the names of its
-   contributors may be used to endorse or promote products derived from
-   this software without specific prior written permission.
+* Neither the name of the vector package developers nor the names of its
```
Comment (HaydenMcT) [inline LICENSE]: This line we might want to leave as copyright holders/developers, right?

File: README.md [hunk 2]
```diff
@@ -1,443 +1,302 @@
-# Fast Sparse Decision Tree Optimization via Reference Ensembles
+# Generalized Optimal Sparse Decision Trees
```
Comment (HaydenMcT) [inline README.md]: This codebase is the implementation for Fast Sparse Decision Tree Optimization via Reference Ensembles. It's also where pypi points now for the currently up-to-date GOSDT code, but I think people should be citing both, especially if they're using depth constraints and threshold guessing, two of the key additions with the Fast Sparse paper. I'd like to have the readme reflect that more clearly, and at the very least to continue citing both papers, not just one (I can take a pass later for this).

File: scripts/gosdt-guesses/README.md
Comment (HaydenMcT) [inline scripts/gosdt-guesses/README.md:1]: If we want to add these, that's great, but we shouldn't push these files to our main public repo until they're ready

File: scripts/gosdt/README.md
Comment (HaydenMcT) [inline scripts/gosdt/README.md:1]: (same as above - I say we don't push these until we can push them without the TODO)

File: src/gosdt/__init__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+"""
+Copyright (c) 2024 Ilias Karimalis. All rights reserved.
```
Comment (HaydenMcT) [inline src/gosdt/__init__.py]: Ilias, are we good to copy over the complete readme copyright info, for consistency? All rights reserved seems incompatible with our 3-clause license.

File: src/gosdt/__init__.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+gosdt: Implementation of General Optimal Sparse Decision Tree
```
Comment (HaydenMcT) [inline src/gosdt/__init__.py]: (and fast sparse reference ensembles!)

File: src/libgosdt/include/configuration.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    /// Dumps the configuration object to a JSON object
+    json to_json() const;
+
+    /// Save and Load the configuration object to a file
+    void save(const std::string &path) const;
+    static Configuration load(const std::string &path);
+
+   public:
+    // #TODO(Ilias): Replace the original configuration with these capitalized/renamed
```
Comment (HaydenMcT) [inline src/libgosdt/include/configuration.hpp:31]: I'm ok leaving this as a todo for the merge, but if this is quickly resolvable now we may as well swap these in first.

**Review by HaydenMcT — COMMENTED** [review]
It's difficult to overstate how much of a benefit this is, thank you again for all of your work on this!
I don't have too many comments since we discussed this already on the private repository; I just flagged a couple superficial spots where I wanted to confirm we're good to submit as is (i.e. we may not want to push some of the scripts readmes until they're fully ready). I also have some thoughts about additions to the readme, though some of these can be done post-merge.

File: .gitignore
Comment (ilias-karimalis) [reply .gitignore:1]: Seems like I incorrectly forgot to add the new `.gitignore` and github workflows file. Adding them back in with a new commit now.

File: LICENSE (hunk 1 above)
Comment (ilias-karimalis) [reply LICENSE]: This seems to be a result of the Python project template this was built on top of. Reversing the change now!

File: src/libgosdt/include/configuration.hpp (hunk 5 above)
Comment (ilias-karimalis) [reply src/libgosdt/include/configuration.hpp:31]: I'm leaving this in for now and will deal with it later!

**Review by HaydenMcT — APPROVED** [review]
Looks great!

This will also close PR #10 , resolving the underlying problem with predictions and removing the need for that PR.

**Pull request merged**